### ISSUE TITLE
deployment to remote hosts breaks after some time

### ISSUE THREAD

**Issue by lUNuXl** [docker/compose#9870]
**Description**

`docker compose` is unable to deploy stack(s) to remote hosts, the hostname appears to be hardcoded in docker(?)

```
bartek@czajniczek ~/Desktop [18]> docker compose -p "project_name" pull
[+] Running 0/0
 ⠿ <<service_name_obfuscated>> Error                                                                                      0.0s
 ⠿ <<service_name_obfuscated>> Error                                                                                 0.0s
 ⠿ <<service_name_obfuscated>> Error                                                                    0.0s
 ⠿ <<service_name_obfuscated>> Error                                                                             0.0s
 ⠿ <<service_name_obfuscated>> Error                                                                                       0.0s
error during connect: Post "http://docker.example.com/v1.40/images/create?fromImage=postgres&tag=11-alpine": command [ssh -l testuser -- <<internal_server_ipv4_address_obfuscated>> docker system dial-stdio] has exited with signal: killed, please make sure the URL is valid, and Docker 18.09 or later is installed on the remote host: stderr=
bartek@czajniczek ~/Desktop [18]> docker context ls
NAME               DESCRIPTION                               DOCKER ENDPOINT               KUBERNETES ENDPOINT   ORCHESTRATOR
company_name-machine *                                             ssh://testuser@<<internal_server_ipv4_address_obfuscated>>                          
default            Current DOCKER_HOST based configuration   unix:///var/run/docker.sock                         swarm
bartek@czajniczek ~/Desktop> 
```

**Steps to reproduce the issue:**
1. `docker context create ssh-box --docker "host=ssh://user@my-box"`
2. `docker context use ssh-box`
4. start using this context to do normal docker things, deploy a stack a few times, tear it down, etc.
5. `docker context ls`
6. `docker compose up`

**Describe the results you received:**

After some time the connection breaks(?) and docker decides that I wanted to call example.com address instead of whatever address I defined in the context.

**Describe the results you expected:**

Seamless communication with the remote docker daemon

**Additional information you deem important (e.g. issue happens only occasionally):**

> issue happens only occasionally

After some time, at first it works, then *something* breaks. I can't tell if it's the remote daemon that does something or if its the client side.

**Output of `docker compose version`:**

```
Docker Compose version 2.10.2
```

**Output of `docker info`:**

```
Client:
 Context:    <<company_name>>-machine
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc., 2.10.2)

Server:
 Containers: 60
  Running: 48
  Paused: 0
  Stopped: 12
 Images: 116
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1062.9.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 31.42GiB
 Name: <<internal_hostname>>
 ID: <<internal>>
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  <<internal_registry>>
  <<<<internal_registry>>
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details:**

Local [client] OS is ArchLinux 64bit and this issue also happens with Debian 11 64bit hosts with official docker package (the one installed from official docker apt repository)

**Comment by nicksieger** [comment]
Could be related to #9448?

**Comment by lUNuXl** [comment]
> Is there anything that makes you think this is related to compose specifically and not any docker command? Are you able to `docker run` or `docker ps` against the remote host after it breaks?

Yes, I've run into this issue a few minutes ago and `docker ps` works, `docker compose pull` doesn't work at all (mentions that cant connect to docker.example.com) and `docker compose down -v` deletes one volume at a time (I have to run this command 4 times to delete 4 containers because it fails after deleting the first one)

**Comment by lUNuXl** [comment]
> Could be related to #9448?

I think this is related, I even followed the advice there about increasing the max sessions and rerites in ssh daemon settings on one of the hosts and this problem still persists.

I didn't have time to test if downgrading docker compose on the client machine fixes anything, I'll probably try doing this later today/tomorrow (that is, if I at all find time to fiddle with it this weekend)

**Comment by lUNuXl** [comment]
Update: I've downgraded docker compose package and everything works now

More information: https://github.com/docker/compose/issues/9448#issuecomment-1265437932

**Comment by AlexZeitler** [comment]
Downgrading to `2.3.3` seems to fix it for me as well. But I'm also getting the warnings described in https://github.com/docker/compose/issues/8544.

**Comment by matt0x6F** [comment]
I'm able to reproduce this in compose 2.16.0. Same symptoms, `docker compose ps` will work *one time*, then it returns a similar error. I checked the SSH logs the connection was terminated by the client. On the other hand `docker ps` shows all the right containers. This is definitely a compose bug.

**Comment by ndeloof** [comment]
@matt0x6F could you please try to reproduce by directly executing the docker-compose CLI plugin ?
`.docker/cli-plugins/docker-compose --debug ps`
(or  `/Applications/Docker.app/Contents/Resources/bin/docker-compose --debug ps` when using Docker Desktop Mac)

`docker compose` uses the exact same Docker client set by the docker CLI, just like `docker ps` does, but maybe this issue happens due to the docker CLI plugin architecture.

**Comment by matt0x6F** [comment]
@ndeloof same output

```
[12:29:03]❯ /home/<user>/.docker/cli-plugins/docker-compose ps 
error during connect: Get "http://docker.example.com/v1.42/containers/82eb731dac747b7d0bd216d00417b0b24baa6da80a654a28f032eb062c8cd90d/json": command [ssh -l matt -- <server ip> docker system dial-stdio] has exited with exit status 255, please make sure the URL is valid, and Docker 18.09 or later is installed on the remote host: stderr=kex_exchange_identification: read: Connection reset by peer
Connection reset by <server ip> port 22
```

**Comment by politician** [comment]
[This](https://github.com/docker/compose/issues/8191#issuecomment-1448646228) fixed it for me!

**Comment by LaXiS96** [comment]
The solution found by @politician works in a non-Windows client environment, but on Windows with Microsoft's OpenSSH implementation you won't get very far: `ControlMaster` is not supported by Microsoft's implementation, which means Windows users pretty much cannot use SSH authentication to the Docker API.

The underlying issue seems to be that compose is spamming separate connections for each docker command, which triggers either the remote host connection limiting or an intermediate firewall (thus `connection reset`).

I notice this behavior consistently when e.g. running `docker compose up -d` for around 20 services, but never when doing the same for a single service.
